### ISSUE TITLE
Shapes a few pixels off when upgrading to fabric.js 4 (beta 12)

### ISSUE THREAD

**Issue by bbernhard** [fabricjs/fabric.js#6323]
Hi, 

I am currently migrating from fabric.js 1.7.19 to fabric.js 4.0 (beta 12) and I am running into a strange problem. For some reason, shapes that were created with fabric.js 1.x are a few pixels off when rendered with fabric.js 4.0. 

My test setup looks like this: 

I have two git branches - ` production` and `development`. On the production branch, fabric.js 1.7.19 is running and on the `development` brunch I am running fabric.js 4.0 beta 12. (I had to change a handful of things on the `development` branch to make fabric.js 4.0 work - but all of those changes where related to the [breaking changes](http://fabricjs.com/changes-introduction). So except of the changes I had to make due to the breaking changes that where introduced in fabric.js 2.0 the code is identical.) 

I am using ` puppeteer` to make a screenshot of both the `production` and the `development` version. 

Here is the screenshot from the `production` version: 

![f2164814-507c-4760-9f30-d05250938506_old](https://user-images.githubusercontent.com/5710631/81415043-2a34d600-9148-11ea-86da-2a4e565452d8.png)

And here is the one from the `development` version:

![f2164814-507c-4760-9f30-d05250938506](https://user-images.githubusercontent.com/5710631/81414971-12f5e880-9148-11ea-9fc1-0a4226ebc6dd.png)

Just by looking at those images I couldn't spot any differences, but then I used ImageMagick's `compare` to compare both images with: 

```compare -metric PAE -highlight-color orange f2164814-507c-4760-9f30-d05250938506.png f2164814-507c-4760-9f30-d05250938506_old.png comparison.png```

And that's the result (the orange lines mark where both images differ): 

![comparison](https://user-images.githubusercontent.com/5710631/81415192-6700cd00-9148-11ea-85fb-ec15b45c8dd6.png)

I also did a manual comparison with gimp (to rule out that it's an ImageMagick issue), but I can see the same there. 

I guess this bug report is probably not sufficient to track down the issue, but before I try to reproduce that with a minimal example I wanted to know if that's something that is already covered by regression tests? At the moment I don't think that the bug is in my implementation (as I really haven't changed much between the `development` and the `production` branch - I just handled the breaking changes), but I also can't rule that out completely. So, if there are already regression tests between versions, I think it's more likely that the issue is in my code :sweat_smile: 

*edit:* I am using `toJSON` for the serialization.

**Comment by bbernhard** [comment]
Ok, I've tracked it down further. It seems that the small offset was introduced in version `1.7.20`. (in the production environment I am running version `1.7.19`). 

In order to verify that was really introduced in version `1.7.20` I created another git branch - directly branched off the `production` branch. The only thing  I changed was the `fabric.min.js` version (from `1.7.19`, which is the version that's running in production, to `1.7.20`). if I do the same tests as above (i.e creating two screenshots with `puppeteer` + comparing the screenshots with `compare` I can see the same difference). 

I just looked at the [diff](https://github.com/fabricjs/fabric.js/compare/v1.7.19...v1.7.20) between version `1.7.19` and `1.7.20`, but as I do not know the code base I have no clue what could be reason for that. Anybody an idea? Any help is really appreciated!

**Comment by bbernhard** [comment]
I think I nailed it down to this commit here: https://github.com/fabricjs/fabric.js/commit/435c5e6f2beb7df82d905aa1671cbe541683d40e

I checked out `v1.7.19` and cherry-picked one commit after another. After every commit that I've cherry-picked I ran my tests. I can see that after adding the above commit, the slight pixel offset appears the first time. In order to confirm that it's really just that commit, I've checked out `v1.7.20` and reverted just that single commit - and it still works. So I am pretty sure that the slight pixel offset was introduced in that commit. 

Does anyone have an idea what could be the reason for that?

**Comment by asturur** [comment]
from what version you are migrating off? 1.x is a bit generic.

**Comment by bbernhard** [comment]
> from what version you are migrating off? 1.x is a bit generic.

It's version `1.7.19`. Sorry, I'll add that to the initial post. As described above, I think it's this commit here: https://github.com/fabricjs/fabric.js/commit/435c5e6f2beb7df82d905aa1671cbe541683d40e

**Comment by asturur** [comment]
So the version  1.7 was where we added caching.
At some point we figured out there was a slight blurryness of the cache because it wasn't drawn as much sharp as possible.

The difference you see is probably less than half pixel and is a rounding for caching positioning.
So in theory before 1.7.20 position was wrong and then it was corrected.

Are you worried for something in particular for this difference?

**Comment by asturur** [comment]
I took your scrennshot, zoomed with the browser, and took a screenshot.
Dev seems indeed less blurry.

dev
![image](https://user-images.githubusercontent.com/1194048/81478464-fc21c580-921d-11ea-99ed-17bd6e5f730e.png)

prod
![image](https://user-images.githubusercontent.com/1194048/81478470-06dc5a80-921e-11ea-80cd-251810b8db78.png)

**Comment by asturur** [comment]
Also if you are just creating annotations around pictures, just disable caching for maximum sharpness.

**Comment by bbernhard** [comment]
Many thanks for taking time to look into this one - very much appreciated!

> Are you worried for something in particular for this difference?

The main reason I was worried a bit, was that I didn't know where this difference was coming from.

At the moment we have ~125k image annotations stored in the [ImageMonkey](https://imagemonkey.io/) dataset - almost all of those annotations were created through the webinterface with fabric.js (can't thank you enough for this wonderful piece of software!). 

And as I am now preparing to upgrade fabric.js from 1.7.19 to (the hopefully soon stable) 4.0, I am double/tripple checking everything for regressions. So I decided to write a simple check which verifies that the old fabric.js version behaves the same after upgrading. As I cannot check every one of those 128k annotations I decided to do some random samples and while doing that I noticed the small difference. If the difference is just due to rounding, than that's totally fine for me - I was just worried that there is maybe a bug hidden somewhere which could result in larger differences under certain conditions. But now I am relieved that it's just due to rounding. Many thanks for taking time to look into this one!

> Also if you are just creating annotations around pictures, just disable caching for maximum sharpness.

many thanks, will check that!
